File: qryDnsTypA.lua
#!/usr/bin/env lua

local dnsServer = "8.8.8.8"
local dnsPort = 53
local qType = 1
local hostName = ""

if #arg < 1 then
  print("\nUsage: "..arg[0].." <HOST> [QRY_TYPE [DNS_SERVER [DNS_PORT]]]\n")
  os.exit()
end
if #arg>3 then dnsPort=arg[4] end
if #arg>2 then dnsServer=arg[3] end
if #arg>1 then qType=arg[2] end
hostName = arg[1]

if qType=="A" then qType=1 end
if qType=="AAAA" then qType=28 end
qType = tonumber(qType)

--print(hostName.."["..qType.."]".."@"..dnsServer..":"..dnsPort)

------------------------------------------------------------------------

local function hex2Str(s)
  -- https://stackoverflow.com/q/65476909
  -- string.fromhex() ==> return (data:gsub("%x%x", function(digits) return string.char(tonumber(digits, 16)) end))
  -- return (s:gsub(".", function(char) return string.format("%02x", char:byte()) end))
  return (s:gsub("%x%x", function(digits) return string.char(tonumber(digits,16)) end))
end
local function str2Hex(s)
  local bytes = {}
  for i=1,s:len()  do
    bytes[#bytes+1] = ('%02x'):format(s:byte(i,i))
  end
  return table.concat(bytes, '')
end

local function bytes2Num(bytes)
  return tonumber(str2Hex(bytes),16)
end
local function bytesBE2Num(bytesBE)
  return bytes2Num(bytesBE:reverse())
end

local function byte2BitsBE(n) -- https://stackoverflow.com/a/9080080
  -- returns a table of bits, least significant first.
  local t={} -- will contain the bits
  while n>0 do
    r = math.fmod(n,2)
    t[#t+1] = string.format("%d", r)
    n = (n-r)/2
  end
  return table.concat(t, '')..string.rep("0",8-#t)
end

local function charAt(s, idx)
  return string.sub(s,idx,idx)
end
local function charAt2BitsBE(s, idx)
  return byte2BitsBE(charAt(s,idx):byte())
end
local function charAt2Bits(s, idx)
  return string.reverse(charAt2BitsBE(s, idx))
end

------------------------------------------------------------------------

local function ipAddrBytes2Str(s)
  local aAddr = ""
  if (#s==4) then
    aAddr = ( charAt(s,1):byte() .. "." .. charAt(s,2):byte()
      .. "." .. charAt(s,3):byte()  .. "." .. charAt(s,4):byte()
    )
  elseif (#s==16) then
    aAddr = ( str2Hex(s:sub(1,2))
      .. ":" .. str2Hex(s:sub(3,4))
      .. ":" .. str2Hex(s:sub(5,6))
      .. ":" .. str2Hex(s:sub(7,8))
      .. ":" .. str2Hex(s:sub(9,10))
      .. ":" .. str2Hex(s:sub(11,12))
      .. ":" .. str2Hex(s:sub(13,14))
      .. ":" .. str2Hex(s:sub(15,16))
    )
  else
    aAddr = str2Hex(s)
  end
  return aAddr
end

------------------------------------------------------------------------

local function getEndOfNameData(rawData, startIdx)
  local idx = startIdx
  while true do
    local sz1 = string.sub(rawData,idx,idx):byte()
    if sz1==0 then
      return idx
    elseif sz1<64 then
      idx = idx+sz1+1
      if idx>#rawData then
        return nil -- data error
      end
    else -- elseif sz1>=192 then
      return idx+1
      -- 4.1.4. Message compression https://datatracker.ietf.org/doc/html/rfc1035
      -- if a reference/pointer is found/used, it must either be the only element
      -- or the last element !! no additional null character after this !!
      -- hence, we should return immediately and ends here
    end
  end
  return nil
end

------------------------------------------------------------------------

local socket = require("socket")

local function qryDns(qName, qType, srv, prt, sck)
  local function getRnd2Bytes()
    -- max value of 2 bytes == 65535 --
    math.randomseed(os.time())
    local t = socket.gettime() -- print(os.time());  print(t);  print(t%60)
    return string.char(math.floor(0.5+((t%10)/10*255)))
      .. string.char(math.random(0, 255))
  end
  local function formatNameData(qName)
    local qData = ""
    for n1,_ in string.gmatch(qName..'.', "([^%.]+)%.") do
      qData = qData .. string.char(#n1) .. n1
    end
    --print(str2Hex(qData))
    return qData
  end
  local function getAnswer(msgData, qType)
    --msgHdrHex = str2Hex(msgData:sub(1,12)) ; print(msgHdrHex)
    -- ID:2bytes; Flags+OpCode+RespCode:2bytes; RecordsCount:4*2bytes
    local byte4 = charAt2Bits(msgData,4)
    local rspCode = tonumber(byte4:sub(5,8), 2)
    local qCount = bytes2Num(msgData:sub(5,6))
    local aCount = bytes2Num(msgData:sub(7,8))
    --print("aCount: "..aCount)
    local byteIdx = 13
    for count = 1,qCount,1 do
      local endOfNameData = getEndOfNameData(msgData, byteIdx)
      byteIdx = endOfNameData+1+4 -- end of q1
    end
    if (qCount>=1 and aCount > 0) then
      for count = 1,aCount,1 do
        local endOfNameData = getEndOfNameData(msgData, byteIdx)
        byteIdx = endOfNameData+1 -- end of a1
        local aType = bytes2Num(string.sub(msgData,byteIdx,byteIdx+1)) -- 1:A(ipv4)|28:AAAA(ipv6)| https://en.wikipedia.org/wiki/List_of_DNS_record_types#Resource_records
        local aClass = bytes2Num(string.sub(msgData,byteIdx+2,byteIdx+3)) -- normally the value 1 for Internet ('IN')
        --print("aType:"..aType.." qType:"..qType)
        byteIdx = byteIdx+8 --
        local aSize = bytes2Num(string.sub(msgData,byteIdx,byteIdx+1))
        byteIdx = byteIdx+2 --
        if (aType==qType) then
          local aData = msgData:sub(byteIdx,byteIdx+aSize-1)
          if (aType==1 or aType==28) then
            return ipAddrBytes2Str(aData)
          else
            return str2Hex(aData)
          end
        end
        byteIdx = byteIdx+aSize
      end
    end
    if (qType==1) then
      return "0.0.0.0"
    elseif (qType==28) then
      return "::"
    else
      return "#"
    end
  end
  sck:settimeout(5);  sck:sendto(
    getRnd2Bytes() .. hex2Str("01000001000000000000")
      .. formatNameData(qName) .. hex2Str("0000")
        .. string.char(qType) .. hex2Str("0001"),
          srv, prt
  )
  local rspData,srcAddrOrErrMsg,srcPort = sck:receivefrom()
  if rspData then
    --print(str2Hex(rspData));  --print(rspData)
    return getAnswer(rspData, qType)
  else
    error("Error: "..tostring(srcAddrOrErrMsg))
  end
end

print( qryDns(hostName, qType, dnsServer, dnsPort, socket.udp()) )

File: dns/qryDnsTypA.py
#!/usr/bin/env lua

import sys
import socket

dnsServer = "8.8.8.8"
dnsPort = 53
qType = 1
hostName = "google.com"

if len(sys.argv) < 2 :
  print("\nUsage: "+sys.argv[0]+" <HOST> [QRY_TYPE [DNS_SERVER [DNS_PORT]]]\n")
  quit()
#
if len(sys.argv)>4 : dnsPort=sys.argv[4] #
if len(sys.argv)>3 : dnsServer=sys.argv[3] #
if len(sys.argv)>2 : qType=sys.argv[2] #
if len(sys.argv)>1 : hostName=sys.argv[1]

if qType=="A" : qType=1 #
if qType=="AAAA" : qType=28 #
qType = int(qType)

##print(hostName+"["+qType+"]"+"@"+dnsServer+":"+dnsPort)

########################################################################

def bytes2Num(bytes) :
  #byteorder is "big" ==> most significant byte is at the beginning ==> little endian
  return int.from_bytes(bytes, byteorder='big', signed=False)
#
def byte2Bits(n) :
  return bin(n).lstrip('0b')
#
def bytes2Hex(bytes) :
  return ''.join('{:02x}'.format(x) for x in bytes)
#

def hex2Str(s) :
  return bytearray.fromhex(s).decode() # s.decode("hex")
#
def str2Hex(s) :
  return bytes2Hex(s.encode("latin-1"))
#

########################################################################

def ipAddrBytes2Str(bytes) :
  aAddr = ""
  if (len(bytes)==4) :
    aAddr = ( str(bytes[0]) + "." + str(bytes[1])
      + "." + str(bytes[2])  + "." + str(bytes[3])
    )
  elif (len(s)==16) :
    aAddr = ( bytes2Hex(bytes[0:1])
      + ":" + bytes2Hex(bytes[2:3])
      + ":" + bytes2Hex(bytes[4:5])
      + ":" + bytes2Hex(bytes[6:7])
      + ":" + bytes2Hex(bytes[8:9])
      + ":" + bytes2Hex(bytes[10:11])
      + ":" + bytes2Hex(bytes[12:13])
      + ":" + bytes2Hex(bytes[14:15])
    )
  else :
    aAddr = bytes2Hex(s)
  #
  return aAddr
#

########################################################################

def findEndOfNameData(rawData, startIdx) :
  idx = startIdx
  while True :
    sz1 = rawData[idx]
    if sz1==0 :
      return idx
    elif sz1<64 :
      idx = idx+sz1+1
      if idx > len(rawData) :
        return nil ## data error
      #
    else : ## elif sz1>=192 :
      return idx+1
      ## 4.1.4. Message compression https://datatracker.ietf.org/doc/html/rfc1035
      ## if a reference/pointer is found/used, it must either be the only element
      ## or the last element !! no additional null character after this !!
      ## hence, we should return immediately and ends here
    #
  #
  return nil
#

########################################################################

def qryDns(qName, qType, srv, prt, sck) :
  def getRnd2Chrs() :
    ## max value of 2 bytes == 65535 ##
    import time;  t = time.time()
    import random;  random.seed()
    return chr(round((t%10)/10*255)) + chr(random.randrange(255))
  #
  def encodeNameData(qName) :
    qData = ""
    for n1 in qName.split('.') :
      qData = qData + chr(len(n1)) + n1
    #
    ##print(str2Hex(qData))
    return qData
  #
  def parseAnswer(msgData, qType) :
    ##msgHdrHex = str2Hex(msgData:sub(1,12)) ; print(msgHdrHex)
    ## ID:2bytes; Flags+OpCode+RespCode:2bytes; RecordsCount:4*2bytes
    byte4 = byte2Bits(msgData[3])
    rspCode = int(byte4[4:], 2)
    qCount = bytes2Num(msgData[4:6])
    aCount = bytes2Num(msgData[6:8])
    ##print("aCount: "+str(aCount))
    byteIdx = 12
    for count in range(qCount) :
      endOfNameData = findEndOfNameData(msgData, byteIdx)
      byteIdx = endOfNameData+1+4 ## end of q1
    #
    if (qCount>=1 and aCount > 0) :
      for count in range(aCount) :
        endOfNameData = findEndOfNameData(msgData, byteIdx)
        byteIdx = endOfNameData+1 ## end of a1
        aType = bytes2Num(msgData[byteIdx:byteIdx+2]) ## 1:A(ipv4)|28:AAAA(ipv6)| https://en.wikipedia.org/wiki/List_of_DNS_record_types#Resource_records
        aClass = bytes2Num(msgData[byteIdx+2:byteIdx+4]) ## normally the value 1 for Internet ('IN')
        ##print("aType:"+str(aType)+" qType:"+str(qType))
        byteIdx = byteIdx+8 ##
        aSize = bytes2Num(msgData[byteIdx:byteIdx+2])
        byteIdx = byteIdx+2 ##
        if (aType==qType) :
          aData = msgData[byteIdx:byteIdx+aSize]
          if (aType==1 or aType==28) :
            return ipAddrBytes2Str(aData)
          else :
            return bytes2Hex(aData)
          #
        #
        byteIdx = byteIdx+aSize
      #
    #
    if (qType==1) :
      return "0.0.0.0"
    elif (qType==28) :
      return "::"
    else :
      return "#"
    #
  #

  qryData = ( getRnd2Chrs() + hex2Str("01000001000000000000") \
    + encodeNameData(qName) + hex2Str("0000") \
    + chr(qType) + hex2Str("0001")
  ).encode("latin-1")
  #print(bytes2Hex(qryData)); print(qryData);

  ##sck.bind(('', 40053)) ## use specific source port !!
  ##sck.connect((srv, prt)); sck.send(qryData) ## this is more standard/proper ??
  sck.sendto(qryData, (srv, prt)) ## this will be more compatible/flexible !!
  #print("sent")

  rspData = sck.recv(1024)
  #print("received")
  if rspData :
    #print(bytes2Hex(rspData)); ##print(rspData)
    return parseAnswer(rspData, qType)
  else :
    print("Error")
  #
#

print( qryDns(hostName, qType, dnsServer, dnsPort, socket.socket(socket.AF_INET,socket.SOCK_DGRAM)) )
